tests/js/server/aql/aql-functions-misc.js

```diff
@@ -733,6 +733,268 @@ function ahuacatlMiscFunctionsTestSuite () {
+          assertEqual(sid._documents[0], "cl");
+        } else {
+          counts = db.cl.count(true);
+          assertEqual(counts[sid._documents[0]], 1);
+        }
+      } finally {
+        db.cl.drop();
+      }
+
+      try {
+        cl = db._create("cl", {numberOfShards:3, shardKeys:["a", "b"]});
+        val = 3.1415926535897932384626433832795028841971693993751058209749445923078164062;
+        d = db.cl.insert({"a":val});
+        sid = db._query('RETURN SHARD_ID("cl", {a:@val})', {val});
+        if (!isCluster) {
+          assertEqual(sid._documents[0], "cl");
+        } else {
+          counts = db.cl.count(true);
+          assertEqual(counts[sid._documents[0]], 1);
+        }
+      } finally {
+        db.cl.drop();
+      }
+
+      try {
+        cl = db._create("cl", {numberOfShards:3, shardKeys:["a", "b"]});
+        vala = 3;
+        valb = "Pi";
+        d = db.cl.insert({"a":vala,"b":valb});
+        sid = db._query('RETURN SHARD_ID("cl", {a:@vala, b:@valb})', {vala, valb});
+        if (!isCluster) {
+          assertEqual(sid._documents[0], "cl");
+        } else {
+          counts = db.cl.count(true);
+          assertEqual(counts[sid._documents[0]], 1);
+        }
+      } finally {
+        db.cl.drop();
+      }
+
+      try {
+        cl = db._create("cl", {numberOfShards:3, shardKeys:["a", "b"]});
+        vala = "Pi";
+        valb = 3.1415926535897932384626433832795028841971693993751058209749445923078164062;
+        d = db.cl.insert({"a":vala,"b":valb});
+        sid = db._query('RETURN SHARD_ID("cl", {a:@vala, b:@valb})', {vala, valb});
+        if (!isCluster) {
+          assertEqual(sid._documents[0], "cl");
+        } else {
+          counts = db.cl.count(true);
+          assertEqual(counts[sid._documents[0]], 1);
+        }
+      } finally {
+        db.cl.drop();
+      }
+
+      try {
+        cl = db._create("cl", {numberOfShards:3, shardKeys:["a", "b"]});
+        vala = 3.1415926535897932384626433832795028841971693993751058209749445923078164062;
+        valb = 3;
+        d = db.cl.insert({"a":vala,"b":valb});
+        sid = db._query('RETURN SHARD_ID("cl", {a:@vala, b:@valb})', {vala, valb});
+        if (!isCluster) {
+          assertEqual(sid._documents[0], "cl");
+        } else {
+          counts = db.cl.count(true);
+          assertEqual(counts[sid._documents[0]], 1);
+        }
+      } finally {
+        db.cl.drop();
+      }
+
+      // check that every document among 1000 went to the predicted shard
+      try {
+        cl = db._create("cl", {numberOfShards:3, shardKeys:["a", "b"]});
+        var docs = [];
+        var i;
+        for (i = 0; i < 1000; ++i) {
+          docs.push({i, doc: db.cl.insert({a:i,b:"foo"+i})});;
+        }
+        i = 0;
+        docs.forEach(function (d) {
+          sid = db._query('RETURN SHARD_ID("cl", {a:@a, b:@b})', {a:i,b:"foo"+i});
+          d = db._query('FOR i in cl FILTER i._key == @d.doc._key RETURN i', {d},
```

Comment:
This test has a relatively high probability to be green in spite of something being wrong. For example, if the `shard` option would not work, then it is still ok, since it finds the one document with the right key. How about the following: Insert the 1000 docs as before, filter those that are on a single shard, then query that shard but without a filter, restricted to the shard. Then compare if you get exactly those docs which are supposed to be on that shard. I think this is not much more complicated but tests more the connection between the "shard" option and the `SHARD_ID` function.

Reply:
done